crates/rune-modules/src/time.rs

```diff
@@ -31,51 +31,334 @@
+
     Ok(module)
 }
 
+/// Sleep for the given [`Duration`].
+///
+/// # Examples
+///
+/// ```rune,no_run
+/// use time::Duration;
+///
+/// let d = Duration::from_secs(10);
+/// time::sleep(d).await;
+/// println!("Surprise!");
+/// ```
+#[rune::function]
+async fn sleep(duration: Duration) {
+    tokio::time::sleep(duration.inner).await;
+}
+
+#[rune::function]
+async fn interval(duration: Duration) -> Interval {
+    Interval {
+        inner: tokio::time::interval(duration.inner),
+    }
+}
+
+// #[rune::function]
+// async fn timeout(duration: Duration, future: F) -> Timeout<F>
+// where
+//     F: Future + rune::compile::Named + CoreTypeOf + MaybeTypeOf + 'static,
+// {
+//     Timeout {
+//         inner: tokio::time::timeout(duration.inner, future).await,
+//     }
+// }
+
+fn module_duration(module: &mut Module) -> Result<(), ContextError> {
+    module.ty::<Duration>()?;
+
+    module.function_meta(Duration::new__meta)?;
+    module.function_meta(Duration::from_secs__meta)?;
+    module.function_meta(Duration::from_millis__meta)?;
+    module.function_meta(Duration::from_micros__meta)?;
+    module.function_meta(Duration::from_nanos__meta)?;
+    module.function_meta(Duration::as_secs_f64__meta)?;
+    module.function_meta(Duration::from_secs_f64__meta)?;
+
+    module
+        .constant("SECOND", Duration::from_secs(1))
+        .build_associated::<Duration>()?
+        .docs(["The duration of one second."])?;
+
+    module
+        .constant("MILLISECOND", Duration::from_millis(1))
+        .build_associated::<Duration>()?
+        .docs(["The duration of one millisecond."])?;
+
+    module
+        .constant("MICROSECOND", Duration::from_micros(1))
+        .build_associated::<Duration>()?
+        .docs(["The duration of one microsecond."])?;
+
+    module
+        .constant("NANOSECOND", Duration::from_nanos(1))
+        .build_associated::<Duration>()?
+        .docs(["The duration of one nanosecond."])?;
+
+    module
+        .constant("ZERO", Duration::from_nanos(0))
+        .build_associated::<Duration>()?
+        .docs(["A duration of zero time."])?;
+
+    module
+        .constant("MAX", Duration::new(u64::MAX, NANOS_PER_SEC - 1))
+        .build_associated::<Duration>()?
+        .docs(["A duration of zero time."])?;
+
+    Ok(())
+}
+
 #[derive(Debug, Clone, Copy, Any)]
 #[rune(item = ::time)]
-struct Duration {
+pub struct Duration {
     inner: tokio::time::Duration,
 }
 
 impl Duration {
-    /// Construct a duration from the given number of seconds.
-    /// 
+    /// Creates a new `Duration` from the specified number of whole seconds and
+    /// additional nanoseconds.
+    ///
+    /// If the number of nanoseconds is greater than 1 billion (the number of
+    /// nanoseconds in a second), then it will carry over into the seconds provided.
+    ///
+    /// # Panics
+    ///
+    /// This constructor will panic if the carry from the nanoseconds overflows
+    /// the seconds counter.
```

Comment:
Rune methods must not actually panic, like a Rust panic since that could allow the script host to panic at runtime.

This must be checked, and the panicking behavior must use `VmResult::panic`.

Reply:
Vm Panics, makes more sense?

Reply:
It does, but the new implementation still panics. You'll have to either use a checked operation when constructing the duration or [re-implement the logic](https://doc.rust-lang.org/nightly/src/core/time.rs.html#211) which is used to ensure that it's never fed values which panic.